Avoid redundant call to CGContextClipToRect in GraphicsContextCG::fillRect when filling with a gradient

mwyrzykowski opened the pull request:
#### 6e0387dd939c98e796bec76d97b4737e88fe63f3
<pre>
Avoid redundant call to CGContextClipToRect in GraphicsContextCG::fillRect when filling with a gradient
<a href="https://bugs.webkit.org/show_bug.cgi?id=276623">https://bugs.webkit.org/show_bug.cgi?id=276623</a>
&lt;radar://131632207&gt;

Reviewed by NOBODY (OOPS!).

Skip CGContextClipToRect call when the rect is completely contains
the current clipping rect. The result of intersecting the clipping
rectangle with a rectangle which completely contains it is equivalent
to a nop.

* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::fillRect):
Avoid save / restore cost in the no-clipping case, since we can
simply apply the inverse of the affine transform which is less
expensive than a save and restore.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6e0387dd939c98e796bec76d97b4737e88fe63f3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58511 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37838 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10995 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62137 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8955 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60640 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45474 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9152 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47368 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6377 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60542 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35418 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50585 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28219 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32168 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7887 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7959 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54119 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8158 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63840 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2424 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8167 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54689 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2432 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50611 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54764 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2046 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33667 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34753 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35837 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34498 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
```diff
@@ -479,13 +479,17 @@ void Recorder::drawFocusRing(const Vector<FloatRect>& rects, float outlineOffset
 {
+    auto& state = currentState();
+    auto& graphicsState = state.state;
+    bool performRectIntersection = graphicsState.fillBrush().gradient() && !graphicsState.dropShadow();
```
Did you notice bugs when drop shadows are present? Did tests detect them? If not, please add a test.

I wonder if there is other graphics context state we should consult.

mwyrzykowski replied:
I did not notice any issues with drop shadows as they would not reach the path where the clipping flag is checked. The flag is not consulted when drop shadows are present (see GraphicsContextCG.cpp:820 - `if (hasDropShadow()) {`), so the goal of the check is to avoid the matrix multiplication when it is not needed.

We could compute this when drop shadows are present, but I don't think it is necessary since we will not use the variable.

mwyrzykowski replied:
EWS found a failure :) 
imported/mozilla/svg/opacity-and-gradient-02.svg

I will take a look

mwyrzykowski committed:
Avoid redundant call to CGContextClipToRect in GraphicsContextCG::fillRect when filling with a gradient
https://bugs.webkit.org/show_bug.cgi?id=276623
<radar://131632207>

Reviewed by NOBODY (OOPS!).

Skip CGContextClipToRect call when the rect is completely contains
the current clipping rect. The result of intersecting the clipping
rectangle with a rectangle which completely contains it is equivalent
to a nop.

* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::fillRect):
Avoid save / restore cost in the no-clipping case, since we can
simply apply the inverse of the affine transform which is less
expensive than a save and restore.

smfr commented on Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
```diff
@@ -832,9 +833,15 @@ void GraphicsContextCG::fillRect(const FloatRect& rect, Gradient& gradient, cons
         CGContextDrawLayerInRect(context, rect, layer.get());
     } else {
-        CGContextClipToRect(context, rect);
+        bool skipClip = getCTM().mapRect(rect).contains(CGContextGetDeviceClipBoundingBox(context));
```
Why is this not `CGContextGetClipBoundingBox()` ?

smfr replied:
And you need to explain, either in a comment or the commit message, why mapping the rect through the current CTM is the right thing to do. Could you compare the unmapped rect with `CGContextGetClipBoundingBox()`? If not, why not?

mwyrzykowski replied:
Indeed those two should be equivalent, I will try that instead

smfr commented on Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
```diff
@@ -832,9 +833,15 @@ void GraphicsContextCG::fillRect(const FloatRect& rect, Gradient& gradient, cons
         CGContextConcatCTM(context, gradientSpaceTransform);
         gradient.paint(*this);
+        if (skipClip)
+            CGContextConcatCTM(context, CGAffineTransformInvert(gradientSpaceTransform));
```
Are you also trying to save the cost of a save/restore in the no-clip case? The commit message doesn't mention that either.

mwyrzykowski replied:
Oh yes that is the point, I will make a note in the commit message
